[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`

StrongerXi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #149792
* #149484
* __->__ #149483
* #149482

This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D71906139](https://our.internmc.facebook.com/intern/diff/D71906139)

StrongerXi committed:
[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`

This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi commented:
@StrongerXi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71906139).

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71906139](https://our.internmc.facebook.com/intern/diff/D71906139)

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71906139](https://our.internmc.facebook.com/intern/diff/D71906139)

[ghstack-poisoned]

StrongerXi commented:
@StrongerXi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71906139).

bdhirsh commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1447,32 +1450,46 @@ def call_function(
+        from .torch_function import TensorWithTFOverrideVariable
+
+        new_func = self.value.__new__
+        if new_func is torch.Tensor.__new__:
```
hmm one question - there should be some handling in dynamo today to directly proxy tensor subclass constructors (I think believe from this PR: https://github.com/pytorch/pytorch/pull/135151/files#diff-4a52059570bb96333d8383ce6a9d01bbb114c5e34aff6028f820899ca39b5a26R517 )

Do you know why that code isn't getting hit, and instead you're forced to manually handle the `__new__` and `__init__` yourself in dynamo?

Part of the reason I ask is that we did find some rare cases where this can happen - if dynamo is forced to run some `SubclassCtr()` code in vanilla cpython, but installs itself on the next frame, then dynamo will in fact directly see a `__new__` call. We ended up deciding to handle this a while ago though by forcing dynamo run the entire constructor call in eager via putting `_disable_dynamo` on subclass constructors: https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_api.py#L244.

StrongerXi replied:
Discussed offline:

What @bdhirsh said above applies to `__torch_dispatch__` style tensor subclass (that defines `__tensor_flattens__`, etc.).

This stack focuses on the other types of tensor subclasses (no `__torch_dispatch__`, optional `__torch_function__`). That's why we have a different code path here (both historically and all the newly added bits in this stack). #149792 makes the distinction more clear in `builder.py`.

StrongerXi commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1447,32 +1450,46 @@ def call_function(
+                    ],
+                )
+        else:
+            # Let Dynamo trace through custom `__new__`
+            var = VariableTracker.build(tx, new_func).call_function(
+                tx, [self] + args, kwargs
+            )
```
A torch-function-style tensor subclass custom `__new__` (as show in tests) typically consists of some `_make_subclass` call to construct the instance, and optional attribute mutations. This PR handles the `_make_subclass` part, and #149482 handles the attribute mutations part, thus we can safely trace through the custom `__new__`.

zou3519 commented on torch/_dynamo/polyfills/tensor.py:
```diff
@@ -0,0 +1,36 @@
+    cls: type[Any], data: torch.Tensor, requires_grad: bool = False, **kwargs: Any
+) -> Any:
+    # This is a rough approximation of `THPVariable_make_subclass`. It should
+    # suffice for most of Dynamo tracing purposes.
```
nit: link url to python_variable.cpp

StrongerXi replied:
Updated.

zou3519 commented on torch/_dynamo/polyfills/tensor.py:
```diff
@@ -0,0 +1,36 @@
+    # suffice for most of Dynamo tracing purposes.
+    assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
+    data = data.detach()
+
```
What's the behavior of torch_dispatch_mode x Dynamo today? Do we graph break?

If we don't then we need to emulate https://github.com/pytorch/pytorch/blob/4271ebdbdc43143215896cd128ca6eb5c1aea17a/torch/csrc/autograd/python_variable.cpp#L615-L616

zou3519 commented on torch/_dynamo/polyfills/tensor.py:
```diff
@@ -0,0 +1,36 @@
+    assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
+    data = data.detach()
+
+    # Avoid unnecessary `requires_grad` mutation, which isn't supported in Dynamo.
+    if data.requires_grad != requires_grad:
+        data.requires_grad = requires_grad
```
uhh, can we just assert requires_grad=False? Dynamo's not good with intermediate tensors that require grad (hence `Tensor.requires_grad_` graph breaking, which is probably what you ran into)

StrongerXi replied:
Hmm but asserting would lead to a graph break as well, and it's probably going to be more confusing than the existing graph break?

zou3519 commented on torch/_dynamo/polyfills/tensor.py:
```diff
@@ -0,0 +1,36 @@
+    # Avoid unnecessary `requires_grad` mutation, which isn't supported in Dynamo.
+    if data.requires_grad != requires_grad:
+        data.requires_grad = requires_grad
+
```
Is it possible to graph break on things we don't support? like the rest of this https://github.com/pytorch/pytorch/blob/4271ebdbdc43143215896cd128ca6eb5c1aea17a/torch/csrc/autograd/python_variable.cpp#L628-L643

(does torch._dynamo.graph_break() in a polyfill work?)

StrongerXi replied:
Isn't this what the assert above handles?
```python
assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
```

zou3519 replied:
What I meant was:
- setting requires_grad=True is probably not OK here. Does the user code require that requires_grad=True is set here? If not, can we graph break if requires_grad=True?
- we are not calling set_python_custom_device, set_python_custom_layout, _change_backend_component_keys in the polyfill. If dispatch_device and dispatch_layout etc are False then this is fine. If the user calls make_subclass with dispatch_device=True then we should probably graph break

StrongerXi replied:
Ah I see. I think we would graph break (as I intended) in these cases:
> If not, can we graph break if requires_grad=True?

> If dispatch_device and dispatch_layout etc are False then this is fine. If the user calls make_subclass with dispatch_device=True then we should probably graph break

With the assert I pasted above, we are even more strict on the second case, i.e., if user explicitly pass in extra kwarg like `dispatch_device` at all, we'd graph break.

For the first case (setting `requires_grad`), if that happens we'd just hit the following and graph break accordingly.
https://github.com/pytorch/pytorch/blob/1d9343b44400718291cb1e75ee54f3489f64718f/torch/_dynamo/variables/builtin.py#L1928-L1934

zou3519 replied:
Fair enough, thanks for walking me through it!

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -796,6 +796,9 @@ def method_as_subclass(self, cls):
                 object(), var, mutation_type_cls=AttributeMutationNew
             )
             return var
+        unimplemented(
```
v2? :P

StrongerXi replied:
Good call I'll update it.

zou3519 reviewed (commented):
this looks reasonable to me. Can we induce graph breaks if someone passes make_subclasses things that we do not support?

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71906139](https://our.internmc.facebook.com/intern/diff/D71906139)

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71906139](https://our.internmc.facebook.com/intern/diff/D71906139)

[ghstack-poisoned]

StrongerXi committed:
Update on "[dynamo] Support `torch.Tensor._make_subclass` and tracing through tensor subclass `__new__`"


This builds off the previous patch in the stack, and fully fixes
https://github.com/huggingface/diffusers/issues/10795.

Essentially, tensor subclass in the issue uses
`torch.Tensor._make_subclass`, which has a pretty simple shallow-copy
plus type change semantics, as far as Dynamo is concerned. So this patch
adds a polyfill for it.

As a result, this allows us to trace through many user-defined `__new__`
in tensor subclass (it's similar to how we trace through user-defined
`__new__` for `UserDefinedClassVariable`), so this patch also faithfully
trace through these `__new__` methods.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71906139](https://our.internmc.facebook.com/intern/diff/D71906139)

[ghstack-poisoned]

StrongerXi commented:
@StrongerXi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71906139).

StrongerXi commented:
@StrongerXi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71906139).
